### DISCUSSION TITLE
Error during spacemake run

### DISCUSSION THREAD

**Discussion by vgulaya** [rajewsky-lab/openst#48] [Q&A]
Hello,

I am trying to use openst tool for analysing my data which were generated by following Sec-scope protocol with MiSeq flow cell. I installed openst and spacemake, and registered my sample under the "seq_scope" run-mode. It looks like everything goes fine, as spacemake writes just before the Error "23 of 25 steps (92%) done". Then in the step 24 rule create_automated_report it generates the following error:
 
Error in `ans[ypos] <- rep(yes, length.out = len)[ypos]`:
! replacement has length zero
Backtrace:
  1. base::print(nhood_plot)
  2. ggplot2:::print.ggplot(nhood_plot)
  4. ggplot2:::ggplot_build.ggplot(x)
  5. layout$setup_panel_params()
  6. ggplot2 (local) setup_panel_params(..., self = self)
     ...
 13. ggplot2:::expand_limits_scale(scale, expansion, limits, coord_limits = coord_limits)
 14. ggplot2:::expand_limits_discrete(...)
 15. ggplot2:::expand_limits_discrete_trans(...)
 16. ggplot2:::expand_limits_continuous_trans(...)
 17. base::ifelse(is.na(coord_limits), limits, coord_limits)

Quitting from lines 340-465 [plot_clusters_umap_puck] (tmplw2luo6q.automated_analysis_create_report.Rmd)
Execution halted
[Sat Jun  8 21:09:13 2024]
Error in rule create_automated_report:
    jobid: 24
    output: projects/MiSeq/processed_data/MiSeq0001/illumina/complete_data/automated_analysis/seq_scope/umi_cutoff_100/MiSeq0001_bottom_tiles_automated_report.html

RuleException:
CalledProcessError in line 723 of /srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/site-packages/spacemake/snakemake/main.smk:
Command 'set +o pipefail; JAVA_TOOL_OPTIONS="-Xmx8g -Xss2560k" ;  Rscript --vanilla -e 'rmarkdown::render("/srv/home/vgul0001/openstAnalysis/MiSeq/.snakemake/scripts/tmplw2luo6q.automated_analysis_create_report.Rmd", output_file="/srv/home/vgul0001/openstAnalysis/MiSeq/projects/MiSeq/processed_data/MiSeq0001/illumina/complete_data/automated_analysis/seq_scope/umi_cutoff_100/MiSeq0001_bottom_tiles_automated_report.html", quiet=TRUE, knit_root_dir = "/srv/home/vgul0001/openstAnalysis/MiSeq", params = list(rmd="/srv/home/vgul0001/openstAnalysis/MiSeq/.snakemake/scripts/tmplw2luo6q.automated_analysis_create_report.Rmd"))'' returned non-zero exit status 1.
  File "/srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/site-packages/snakemake/executors/__init__.py", line 2330, in run_wrapper
  File "/srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/site-packages/spacemake/snakemake/main.smk", line 723, in __rule_create_automated_report
  File "/srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/site-packages/snakemake/executors/__init__.py", line 569, in _callback
  File "/srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/concurrent/futures/thread.py", line 58, in run
  File "/srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/site-packages/snakemake/executors/__init__.py", line 555, in cached_or_run
  File "/srv/home/vgul0001/.conda/envs/vgu/envs/spacemake/lib/python3.10/site-packages/snakemake/executors/__init__.py", line 2362, in run_wrapper
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /srv/home/vgul0001/openstAnalysis/MiSeq/.snakemake/log/2024-06-08T171839.255741.snakemake.log
ERROR: SpacemakeError
an error occurred while snakemake() ran

I've tried to search the first error "`ans[ypos] <- rep(yes, length.out = len)[ypos]`: ! replacement has length zero", but the solutions proposed include 
(1) changing gene names to Ensemble or Entrez IDs to provide unique row names (this I already have - all genes are annotated as Ensemble IDs)   
(2) changing Indexes for R-loops as there is a difference in indexing between R and python, but in the script automated_analysis_create_report.Rmd it is very difficult to understand where exactly could be the error.
 
I would really appreciate any help and suggestions on how to deal with this issue.

**Comment by danilexn** [comment]
Hello!

thanks for the feedback.

We have seen this error when there are no coordinates for a puck (it is empty), or when the coordinate limits are not properly set. Can you please share your puck barcode files (or the one for the puck where this rule fails) so we can try to reproduce the issue? Also, if you open the h5ad file or csv summary for that puck (inside the dge folder), are there cells/spots, or are these files empty?

Anyway, we are working on a Python implementation of QC reports (see https://github.com/rajewsky-lab/spacemake/pull/77). This means that most  improvements/bugfixes to QC report will be implemented directly in this new branch, instead of the current R-based version.

Best,
Dani

**Reply by vgulaya** [reply]
Hello!

Thank you for the fast reply. Here is the link for puck_file (https://drive.google.com/file/d/1MSczXnMTUqrhQ_B7X1-kobgl-3c92Z51/view?usp=drive_link). I need to mention that I merged 19 tiles of MiSeq flow cell in this document, because when I was adding my sample to spacemake, it refused to accept multiple files with *.tsv or *.txt specification, so I merged all barcode files in one: 
awk '(NR == 1) || (FNR > 1)' *.tsv > bottom_tiles.tsv

h5ad file (dge.exon.polyA_adapter_trimmed.spatial_beads.mesh_10_hexagon_bottom_tiles.h5ad) is not empty, when I read it with scanpy it says that the object with n_obs × n_vars = 2896 × 13727. 

The csv file is also not empty (dge.exon.polyA_adapter_trimmed.spatial_beads.mesh_10_hexagon_bottom_tiles.obs.csv) - table of 2896 rows × 14 columns

**Comment by danilexn** [comment]
From what I see, the error occurs when plotting the neighborhood matrix. It could also be that there are no clusters (just 1) at the lowest clustering resolution (0.4), can you please check the neighborhood matrix inside the relevant automated analysis folder? Thanks!

**Reply by vgulaya** [reply]
Thank you for suggestion! What file this is exactly? I looked in MiSeq0001_bottom_tiles_obs_df.csv it indeed has a column 'leiden_0.4' which contains only zeroes. Could it be the reason of the error? 
I also looked in the file MiSeq0001_bottom_tiles_nhood_enrichment.csv, it does not have a resolution 0.4 starts with 0.6. The first row look like this:	
    cluster_a   cluster_b	      zscore	               resolution
1	   0	        0	      -0.06489792397308976	     0.6

best,
Valeriia
